### ISSUE TITLE
bug: spinning at 100% CPU (bisected)

### ISSUE THREAD

**Issue by ralfbrown** [darktable-org/darktable#7229]
**Describe the bug**
In the most recent revisions, entering darkroom causes dt to start using 100% CPU even if otherwise entirely idle.  100% CPU use continues even after leaving darkroom.

**To Reproduce**
1. Start dt, wait until fully loaded.  Observe that "top" or similar show minimal or no CPU usage
2. Open any image in darkroom, even one that's never been opened before
3. Observe "top" or similar showing dt now using 100% CPU even while not doing anything
4. Switch back to lighttable
5. Observe that dt continues to use 100% CPU

**Expected behavior**
dt should not be pegging a core when not actually doing anything.

**Platform (please complete the following information):**
 - Darktable Version: 3.3.0+2028
 - OS: linux
 - no OpenCL
 - Compiler: GCC 8.4

**Additional context**
Initial investigation made it look like the gtk main loop was endlessly spinning.

git bisect reports the first bad commit as
```
commit ef3b6d0eb074170efa151e1df858735b0f7ece80
Author: Diederik ter Rahe <dterrahe@yahoo.com>
Date:   Sun Dec 6 18:52:07 2020 +0000
 
    remove widget redraw triggered by picker when leaving darkroom
```
Ping @dterrahe

**Comment by ptilopteri** [comment]
openSUSE Tumbleweed 20201205
NVIDIA GF106 [GeForce GTS 450], 390.138
darktable 3.3.0+2031~g552b2d3f4
bit depth is 64 bit
normal build
SSE2 optimized codepath enabled
OpenMP support enabled
OpenCL support enabled
Lua support enabled, API version 6.1.0
Colord support enabled
gPhoto2 support enabled
GraphicsMagick support enabled
ImageMagick support disabled
OpenEXR support enabled
OpenCL on
i7 12-core 36GB
5.9.12-1-default

I do not have excessive cpu usage, and haven't noticed such in several weeks
and I build master several time each day.
and I test from the control-line

**Comment by ralfbrown** [comment]
Further investigation led me to try the following patch, which eliminates the spinning but might break what dterrahe's commit was addresssing.
```
static int _widget_queue_draw(void *widget)
{
  gtk_widget_queue_draw((GtkWidget*)widget);
  return FALSE;
}

void dt_control_queue_redraw_widget(GtkWidget *widget)
{
  if(dt_control_running())
  {
    g_idle_add(_widget_queue_draw, (void*)widget);
  }
}
```
(wrapping gtk_widget_queue_draw in a function that returns FALSE)

In looking at the Gtk documentation, the function passed to g_idle_add needs to return a boolean, but gtk_widget_queue_draw is void, so you get whatever happens to be in AX at the end of the function as the "return value".  Gtk de-queues the function if it returns FALSE, but keeps calling it if it returns TRUE, i.e. anything but 0 in AX.

**Comment by TurboGit** [comment]
@ralfbrown : this looks ok to me, let's wait for @dterrahe feedback.

**Comment by dterrahe** [comment]
I'll admit to having been "inspired" to directly call `gtk_widget_queue_draw` from `g_idle_add` by a post on stackoverflow without then verifying that it was correct usage. So many thanks @ralfbrown for catching and solving this. Your solution should work fine with the real fix in #7215, calling `g_idle_remove_by_data(module->widget)` before destroying the module, because that is only dependent on the _data_ pointer fed into the function, not the return value.

Would you like to go ahead and submit a PR for this or shall I pick that up (since it is my mess to clean)?

**Comment by dterrahe** [comment]
done
